packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -997,3 +959,1336 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+    // OverlayState's build method always returns a RenderObjectWidget _Theatre,
+    // so it's safe to assume that state equality implies render object equality.
+    if (oldWidget.overlayState != widget.overlayState) {
+      assert(oldWidget.entry == widget.entry);
+      theatre = context.findAncestorRenderObjectOfType<_RenderTheatre>()!;
+      _cachedOverlayEntryTopChildModel?._dispose();
+      _cachedOverlayEntryTopChildModel = null;
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.entry._overlayStateMounted.value = null;
+    widget.entry._didUnmount();
+    _cachedOverlayEntryTopChildModel?._dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return TickerMode(
+      enabled: widget.tickerEnabled,
+      child: _RenderTheatreMarker(
+        theatre: theatre,
+        overlayEntryWidgetState: this,
+        child: widget.entry.builder(context),
+      ),
+    );
+  }
+
+  void _markNeedsBuild() {
+    setState(() { /* the state that changed is in the builder */ });
+  }
+}
+
+
+// * OverlayPortal Implementation
+//  OverlayPortal is inspired by the
+//  [flutter_portal](https://pub.dev/packages/flutter_portal) package.
+//
+// ** RenderObject hierarchy
+// The widget works by inserting its overlay child's render subtree directly
+// under [Overlay]'s render object (_RenderTheatre).
+// https://user-images.githubusercontent.com/31859944/171971838-62ed3975-4b5d-4733-a9c9-f79e263b8fcc.jpg
+//
+// To ensure the overlay child render subtree does not do layout twice, the
+// subtree must only perform layout after both its _RenderTheatre and the
+// [OverlayPortal]'s render object (_RenderLayoutSurrogateProxyBox) have
+// finished layout. This is handled by _RenderDeferredLayoutBox.
+//
+// ** Z-Index of an overlay child
+// [_OverlayLocation] is a (currently private) interface that allows an
+// [OverlayPortal] to insert its overlay child into a specific [Overlay], as
+// well as specifying the paint order between the overlay child and other
+// children of the _RenderTheatre.
+//
+// Since [OverlayPortal] is only allowed to target ancestor [Overlay]s
+// (_RenderTheatre must finish doing layout before _RenderDeferredLayoutBox),
+// the _RenderTheatre should typically be acquired using an [InheritedWidget]
+// (currently, _RenderTheatreMarker) in case the [OverlayPortal] gets
+// reparented.
+
+/// A class to show, hide and bring to top an [OverlayPortal]'s overlay child
+/// in the target [Overlay].
+///
+/// A [OverlayPortalController] can only be given to at most one [OverlayPortal]
+/// at a time. When an [OverlayPortalController] is moved from one
+/// [OverlayPortal] to another, its [isShowing] state does not carry over.
+///
+/// The [show] and [hide] methods typically should not be called when the widget
+/// tree is being rebuilt. However it's allowed to call [show] or [hide] when
+/// the associated [OverlayPortal] widget is yet to be inflated into the widget
+/// tree.
```

Comment:
Maybe this doc could be a little more explicit about when it is OK to call show. Can I call this during `initState` `didUpdateWidget` `didCHangeDependencies` or only e.g. in action handlers between frames?